### PR TITLE
Add a way for specifying sample requests to prepopulate debug forms i…

### PR DESCRIPTION
**Pull request by anuraaga**
…n the doc service. Users provide a list of _args objects to DocService which will be used for matching methods.

```java
List<? extends TBase<?, ?>> sampleRequests = Arrays.asList(
  new hello_args().setName("world"),
  new monster_args().setFood("cookies").setColor("blue"));
...
virtualHostBuilder.serviceUnder("/docs", new DocService(sampleRequests))
```

For #47

### CODE REVIEW

**Comment by anuraaga** [comment]
Hmm - I can't repro the build error even with mvn clean test on my JDK 1.8.0_45. Something fixed since the java version used by travis?

**Comment by anuraaga** [comment]
https://github.com/travis-ci/travis-ci/issues/3259 is tracking Travis CI's java version update, including workarounds we may need for forcing an upgrade in our environment. Trustin, any suggestion?

File: src/main/java/com/linecorp/armeria/server/docs/FunctionInfo.java [hunk 1]
```diff
@@ -45,13 +50,30 @@ static FunctionInfo of(Method method) throws ClassNotFoundException {
         final ClassLoader classLoader = serviceClass.getClassLoader();
 
         @SuppressWarnings("unchecked")
+        Class<? extends TBase<?, ?>> argsClass = (Class<? extends TBase<?, ?>>) Class.forName(
+                serviceName + '$' + methodName + "_args");
```
Comment (delegacy) [inline src/main/java/com/linecorp/armeria/server/docs/FunctionInfo.java]: Use `Class.forName(String name, boolean initialize, ClassLoader loader)` like elsewhere?

File: src/main/java/com/linecorp/armeria/server/docs/FunctionInfo.java [hunk 2]
```diff
@@ -59,12 +81,15 @@ static FunctionInfo of(Method method) throws ClassNotFoundException {
...
     private FunctionInfo(String name,
-                         Class<? extends TBase> argsClass,
-                         Class<? extends TBase> resultClass,
-                         Class<? extends TException>[] exceptionClasses) {
+                         Class<? extends TBase<?, ?>> argsClass,
+                         Class<? extends TBase<?, ?>> resultClass,
+                         Class<? extends TException>[] exceptionClasses,
+                         String sampleRequestJson) {
         this.name = requireNonNull(name, "name");
+        this.sampleRequestJson = requireNonNull(sampleRequestJson);
```
Comment (delegacy) [inline src/main/java/com/linecorp/armeria/server/docs/FunctionInfo.java]: `requireNonNull(sampleRequestJson, "sampleRequestJson")`?

File: src/test/java/com/linecorp/armeria/server/docs/SpecificationTest.java [hunk 3]
```diff
@@ -36,8 +36,12 @@
 import com.linecorp.armeria.server.thrift.ThriftService;
 import com.linecorp.armeria.service.test.thrift.main.FooService;
 import com.linecorp.armeria.service.test.thrift.main.HelloService;
+import com.linecorp.armeria.service.test.thrift.main.HelloService.hello_args;
 
 public class SpecificationTest {
+
+    private static final hello_args SAMPLE_HELLO = new hello_args().setName("sample user");
```
Comment (delegacy) [inline src/test/java/com/linecorp/armeria/server/docs/SpecificationTest.java]: Not used?

**Comment by delegacy** [comment]
Handy feature! :smile: 

Left some minor comments.

File: src/main/java/com/linecorp/armeria/server/docs/DocService.java [hunk 4]
```diff
@@ -44,13 +50,26 @@
...
     public DocService() {
+        this(Collections.emptyList());
+    }
+
+    /**
+     * Creates a new instance, prepopulating debug forms with the provided {@code sampleRequests}.
+     * {@code sampleRequests} should be a list of Thrift argument objects for methods that should be
+     * prepopulated (e.g., a populated hello_args object for the hello method on HelloService).
+     */
+    public DocService(Collection<? extends TBase<?, ?>> sampleRequests) {
```
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/docs/DocService.java]: - `Iterable`?
- Could we also add varargs version?

File: src/main/java/com/linecorp/armeria/server/docs/FunctionInfo.java [hunk 5]
```diff
@@ -119,6 +144,11 @@ public TypeInfo returnType() {
         return exceptions;
     }
 
+    @JsonProperty
+    public String sampleRequestJson() {
+        return sampleRequestJson;
```
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/docs/FunctionInfo.java]: sampleJsonRequest or sampleRequestAsJson?

File: src/main/resources/com/linecorp/armeria/server/docs/assets/js/armeria.js [hunk 6]
```diff
@@ -94,7 +94,7 @@ $(function () {
       'function': functionInfo
     }));
 
-    var debugText = functionContainer.find('.debug-textarea');
+    var debugText = functionContainer.find('.debug-textarea').val(functionInfo.sampleRequestJson);
```
Comment (trustin) [inline src/main/resources/com/linecorp/armeria/server/docs/assets/js/armeria.js]: sampleJsonRequest?

File: src/test/java/com/linecorp/armeria/server/docs/SpecificationTest.java [hunk 7]
```diff
@@ -50,7 +54,8 @@ public void servicesTest() throws Exception {
                                 new VirtualHostBuilder().build(),
                                 PathMapping.ofExact("/foo"),
                                 ThriftService.ofFormats(mock(FooService.AsyncIface.class),
-                                                  SerializationFormat.THRIFT_COMPACT))));
+                                                        SerializationFormat.THRIFT_COMPACT))),
+                                                ImmutableMap.of());
```
Comment (trustin) [inline src/test/java/com/linecorp/armeria/server/docs/SpecificationTest.java]: Collections.emptyMap()

**Comment by anuraaga** [comment]
Thanks - addressed comments

**Comment by delegacy** [comment]
:shipit:

**Pull request merged**